[Kyh196201] 6주차 레스토랑 상세 페이지 구현하기

kyh196201 opened the pull request:
트레이너님 안녕하세요!!

간단한 환경설정이랑 HomePage, NotFoundPage까지 라우팅 적용했고, 테스트도 작성했습니다.

피드백 부탁드립니다 🙂 그리고 이번 한 주도 잘 부탁드립니다 ㅎㅎ!

kyh196201 committed:
React-router-dom 모듈 설치

kyh196201 committed:
Webpack 옵션 수정

kyh196201 committed:
App 테스트 수정

- 헤더가 렌더링되었는지에 대한 테스트를 작성했습니다.

kyh196201 committed:
App 테스트 통과

- header를 구현했습니다.

kyh196201 committed:
Header 테스트 추가

kyh196201 committed:
Header 테스트 통과

kyh196201 committed:
App 컴포넌트 리팩토링

- App 컴포넌트 내부의 header 영역을 Header 컴포넌트로 대체했습니다.

kyh196201 committed:
App 테스트 수정

- Home Page를 렌더링하는 것에 대한 테스트를 추가했습니다.

kyh196201 committed:
App 테스트 통과

- Home Page 레이아웃을 작성했습니다.

kyh196201 committed:
HomePage 테스트 추가

kyh196201 committed:
HomePage 테스트 통과

- HomePage 레이아웃을 작성했습니다.

kyh196201 committed:
HomePage 리팩토링

- 링크 목록에 대한 변수를 선언하고, 테스트 내에서도 import해서 사용하도록, 수정했습니다.

kyh196201 committed:
App 컴포넌트 리팩토링

- HomePage에 대한 레이아웃을 HomePage 컴포넌트로 대체했습니다.

kyh196201 committed:
React Router 적용

- App 컴포넌트에 루트 경로에 해당하는 라우터를 설정했습니다.

kyh196201 committed:
App 테스트 수정

- 정의되지 않은 url로 접근했을때 NotFound를 렌더링하는지에 대한 테스트를 추가했습니다.

kyh196201 committed:
App 테스트 통과

- App 컴포넌트 내부에 NotFoundPage 컴포넌트를 구현했습니다.

passwd10 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,43 @@
+      <MemoryRouter>
+        <App />
+      </MemoryRouter>
+    ));
```
반복되는 패턴이 보이네요. 어떻게하면 반복을 최소화할 수 있을까요?

kyh196201 replied:
렌더 함수를 만들어서 중복을 없앴습니다! 피드백 감사합니다 🙂

passwd10 commented on src/pages/HomePage.test.jsx:
```diff
@@ -0,0 +1,21 @@
+
+    ROUTES.forEach((route) => {
+      expect(getByText(route)).toBeInTheDocument();
+    });
```
시나리오를 검증할 수 있는 테스트를 구현해주시면 좋을 것 같아요. 지금 구현하신 테스트는 link 가 아니어도 통과할 수 있는 테스트인 것 같아요!

kyh196201 replied:
`getByText(route)`를 통해 찾은 엘리먼트가 `href` 속성을 가지고 있는지 테스트하는 코드로 수정했습니다!
더 좋은 방법이 있을까요??

passwd10 reviewed (commented):
TDD 사이클에 맞춰 과제를 잘 진행하고 계신 것 같아요!  👍 
다른 페이지도 빠르게 구현해보시고 반복을 통해 숙련도를 높이는 훈련을 하면 좋을 것 같아요 :)

kyh196201 committed:
NotFoundPage 테스트 추가

kyh196201 committed:
NotFoundPage 테스트 통과

kyh196201 committed:
App 컴포넌트 리팩토링

- 내부에 선언한 NotFoundPage 컴포넌트를 제거했습니다.

kyh196201 committed:
App 테스트 수정

- renderApp 함수를 선언해서 반복되는 render를 제거했습니다.
- about 경로로 접근했을 경우에 대한 테스트를 추가했습니다.

kyh196201 committed:
App 테스트 통과

- App 내부에 AboutPage 컴포넌트를 구현했습니다.
- Page 컴포넌트마다 불필요한 div 엘리먼트를 줄이기 위해서 Routes를 div 태그로 감쌌습니다.

kyh196201 committed:
AboutPage 테스트 추가

kyh196201 committed:
AboutPage 테스트 통과

kyh196201 committed:
NotFoundPage 수정

- 불필요한 변수 ROUTES를 제거했습니다.

kyh196201 committed:
App 테스트 통과

- AboutPage를 별도의 컴포넌트로 분리했습니다.

kyh196201 committed:
HomePage 테스트 수정

- a 태그가 올바르게 렌더링되었는지 확인하도록 테스트를 수정했습니다.

kyh196201 committed:
HomePage 테스트 수정

- useHref() 에러가 발생해서 MemoryRouter로 래핑해서 렌더링하도록 수정했습니다.

kyh196201 committed:
HomePage 테스트 통과

- 라우터를 이동할 수 있는 링크를 구현했습니다.

kyh196201 committed:
App 컴포넌트 테스트 수정

- /restaurants 경로로 접근했을 경우에 대한 테스트 코드를 추가했습니다.

kyh196201 committed:
App 테스트 오타 수정

kyh196201 committed:
App 테스트 통과

- App 테스트 내부에 RestaurantsPage 컴포넌트를 구현했습니다.

kyh196201 committed:
RestaurantsPage 테스트 추가

kyh196201 committed:
RestaurantsPage 테스트 통과

kyh196201 committed:
App 리팩토링

- RestaurantsPage를 별도의 컴포넌트로 분리했습니다.

kyh196201 commented:
트레이너님 피드백 주신 부분 수정완료했고, 라우팅까지 테스트 완료했습니다. 🙂

내일부터는 기존에 있는 코드를 이용해서 빠르게 기능을 구현하려구요 !! 숙달될때까지 반복하겠습니다 ㅎㅎ!

그리고 질문이 있습니다.

[위에 코멘트 남겨주신 부분](https://github.com/CodeSoom/react-week6-assignment-1/pull/98#discussion_r919076236) 수정했는데, 맞게 수정한건지 잘 모르겠네요!

passwd10 commented on src/pages/HomePage.test.jsx:
```diff
@@ -15,7 +15,7 @@ describe('<HomePage />', () => {
 
     ROUTES.forEach((route) => {
-      expect(getByText(route)).toBeInTheDocument();
+      expect(getByText(route)).toHaveAttribute('href');
```
`toHaveAttribute`의 두번째 인자로 path 를 넣어주면 조금 더 명확해질 것 같아요.

passwd10 commented on src/pages/RestaurantsPage.jsx:
```diff
@@ -0,0 +1,15 @@
+  return (
+    <ul>
+      {
+        REGIONS.map((region) => (
```
region을 구조분해할당해서 표현할수도 있어요!

```javascript
REGIONS.map(({ name, id }) => (
  <li key={id}>
    {name}
  </li>
))
```

kyh196201 replied:
이 부분은 아직 수정하지 못했습니다 ㅠ
기능 구현 후 최종적으로 리팩토링 진행하겠습니다 감사합니다!

passwd10 reviewed (commented):
네 과제의 의도에 맞게 잘 진행하고 계신 것 같아요! 👍 
하나의 커밋에는 하나의 기능만 들어갈 수 있도록 작업하는게 좋아요. 커밋을 작은 단위로 가져갈수록 히스토리 파악이 수월해지거든요! 지금은 말씀하신대로 TDD 에 익숙해지기 위해 커밋을 테스트와 기능으로 나누어서 작업하고 계신데, TDD가 익숙해지면 하나의 커밋에서 테스트와 기능을 포함시켜서 작업하기도해요

CI를 통과시킬 수 있도록 나머지 테스트도 작성해보세요! (E2E 테스트에서 통과를 못하고 있네요)

kyh196201 committed:
App 테스트 수정

kyh196201 committed:
RestaurantsPage 테스트 수정

kyh196201 committed:
RestaurantsPage 테스트 통과

kyh196201 committed:
RestaurantsPage 초기 데이터 요청

- RestaurantsPage가 처음 렌더링될 때 loadInitialData 액션을 실행하는지에 대한 테스트를 추가했습니다.
- RestaurantsPage가 처음 렌더링될 때 loadInitialData를 실행했습니다.

kyh196201 committed:
given2 모듈 설치

kyh196201 committed:
given2 모듈 환경설정

kyh196201 committed:
RestaurantsPage 테스트 리팩토링

- given2 모듈을 이용해서 중복되는 useSelector 구현을 제거했습니다.

passwd10 commented on src/App.test.jsx:
```diff
@@ -39,12 +45,20 @@ describe('<App />', () => {
+        categories: CATEGORIES,
+        restaurants: RESTAURANTS,
+      }));
+    });
```
`dispatch`, `useDispatch`, `useSelector` 는 테스트내에서 반복적으로 사용될 확률이 높기때문에 최상단 `describe` 하위에서 한번만 선언해서 재사용하는편이에요

passwd10 commented on src/pages/RestaurantsPage.test.jsx:
```diff
@@ -14,6 +16,16 @@ jest.mock('react-redux');
+    regions: given.regions,
+    categories: given.categories,
+    restaurants: given.restaurants,
+  }));
```
given 라이브러리를 적절하게 사용하셨어요 👍

kyh196201 replied:
감사합니다 :-)
저번 과제 때 알려주셔서 정말 유용하게 사용하고 있습니다!!

kyh196201 committed:
mock service worker 모듈 설치

kyh196201 committed:
api 모킹

- msw 라이브러리를 이용해서 api를 모킹하기 위한 작업을 진행했습니다.

kyh196201 committed:
HomePage 테스트 수정

- href를 가지고 있는 것 뿐아니라 href의 값도 일치하는지 확인하도록 테스트 코드를 수정했습니다.

kyh196201 committed:
RestaurantsContainer 상세 페이지로 이동할 수 있는 링크 구현

- RestaurantsContainer에서 레스토랑 상세 페이지로 이동할 수 있는 링크를 구현했습니다.
- RestaurantsContainer에 링크가 렌더링되었는지 확인하는 테스트 코드를 추가했습니다.
- RestaurantsContainer, RestaurantsPage 테스트에서 Link 컴포넌트를 사용할 수 있도록
MemoryRouter를 추가했습니다.

kyh196201 committed:
RestaurantsContainer 레스토랑 상세 링크 수정

- 레스토랑 각 아이디를 문자열로 변경해서 Link의 to props로 전달했습니다.

kyh196201 committed:
App 테스트 리팩토링

- useDispatch와 dispatch mock 함수를 최상위 describe에서 선언했습니다.

kyh196201 committed:
RestaurantsContainer 상세 페이지 링크 수정

kyh196201 committed:
RestaurantDetailPage 컴포넌트 및 테스트 구현

- App 테스트에서 레스토랑 상세 페이지 경로로 접근했을 때에 대한 테스를 추가했습니다.
- RestaurantDetailPage에서 레스토랑 정보가 있을 경우에 대한 테스트를 추가했습니다.
- 테스트 용으로 사용하기 위해 restaurant fixture를 추가했습니다.

kyh196201 committed:
RestaurantDetailContainer 컴포넌트 구현 및 테스트 작성

kyh196201 committed:
RestaurantDetailPage 컴포넌트 리팩토링

- UI에 관련된 부분을 RestaurantDetailContainer를 이용해서 대체했습니다.

kyh196201 committed:
App 테스트 수정

- 레스토랑 상세 페이지 테스트 중에서 state에 있는 restaurant를 참조하지 못했습니다.
그래서 beforeEach에서 useSelector를 구현하도록 수정했습니다.

kyh196201 committed:
RestaurantDetailContainer loading 중일때 화면 구현

- loading 중일 때에 대한 테스트와 화면을 구현했습니다.

kyh196201 committed:
fetchRestaurant 서비스 및 mock 함수 구현

- api 요청을 위한 fetchRestaurant 함수와 테스트를 위한 mock 함수를 구현했습니다.

kyh196201 commented:
트레이너님 피드백 주신 부분 일부 수정했고, 레스토랑 상세 페이지도 일부 구현했습니다.

급하게 구현하느라, 오타 및 잘못된 테스트 명이 보이는데 리팩토링하면서 전부 수정하겠습니다 😂

질문이 있습니다.

1. 레스토랑 상세 페이지 경로를 설정하는게 많이 고민됐습니다.
결국 api url이랑 동일하게 `/restaurants/restaurantId`로 설정했는데,
리액트 라우터에서 `nested route`를 사용할 때 이런식으로 많이 표현하더라구요.
저는 상세 페이지를 `/restaurants`의 자식 라우터로 설정하지 않았는데, 잘못된 url 설정인가요??

2. 레스토랑 상세 페이지에 궁금한 점이 있습니다.

상세 페이지에 대한 명세를 이렇게 작성했는데, 

```
- 로딩중일 경우
 - 로딩 메시지를 보여준다.

- 레스토랑 정보가 있을 경우
 - 레스토랑 정보를 보여준다.

- 에러가 발생했을 경우
 - 에러 메시지를 보여준다.
```

최초에 접근했을 때 로딩 메시지가 나오기 위해서 `loading`의 초기 값을 `true`로 설정하는게 좋을지
아니면 `loadRestaurant` 액션에서 `loading`의 값을 `true`로 변경해주는게 좋을지 궁금합니다.

kyh196201 committed:
리덕스 초기 상태 값 수정

- 레스토랑 상세 정보를 위한 restaurant, 상세 정보 로딩을 위한 loading을 추가했습니다.

passwd10 commented on package.json:
```diff
@@ -48,5 +48,8 @@
+    "workerDirectory": "public"
   }
-}
+}
```
github에서 `No newline at end of file` 경고를 띄워주고 있네요. 파일의 마지막줄에 한칸을 띄어줘야 해결이 될텐데 왜 이런 경고가 뜨는지 찾아보시면 좋을 것 같아요!

passwd10 commented on src/containers/RestaurantDetailContainer.jsx:
```diff
@@ -0,0 +1,33 @@
+      <p>{address}</p>
+      <h3>메뉴</h3>
+
+      {menu ? (
+        <ul>
+          {
+            menu.map((item) => (
+              <li key={item}>
+                {item}
+              </li>
+            ))
+          }
+        </ul>
+      ) : null}
```
```javascript
{menu && (
  <ul>
    {menu.map((item) => (
        <li key={item}>
          {item}
        </li>
      ))}
  </ul>
)}
```

혹은 

```javascript
  <ul>
    {menu?.map((item) => (
        <li key={item}>
          {item}
        </li>
      ))}
  </ul>
```
이렇게 표현할 수 있을 것 같아요!

kyh196201 replied:
첫 번째 방법 리액트에서 많이 본 코드인데 잊고있었네요 ㅎㅎ 피드백 감사합니다 🙌🙌

passwd10 commented on src/App.jsx:
```diff
@@ -1,32 +1,27 @@
+        <Routes>
+          <Route path="/" element={<HomePage />} />
+          <Route path="/about" element={<AboutPage />} />
+          <Route path="/restaurants" element={<RestaurantsPage />} />
+          <Route path="/restaurants/:restaurantId" element={<RestaurantDetailPage />} />
```
```javascript
<Route path="/restaurants" element={<RestaurantsPage />}>
  <Route path=":restaurantId" element={<RestaurantDetailPage />} />
</Route>
```

이렇게 표현하면 계층구조 파악이 더 쉬워질 것 같아요

passwd10 commented on src/pages/HomePage.jsx:
```diff
@@ -0,0 +1,23 @@
+        {
+          ROUTES.map((route) => (
+            <li key={route}>
+              <Link to={`/${route}`}>
```
App에서 route 설정을 모두 소문자로 하셨으니 여기서도 소문자로 Link 를 걸어주시면 좋을 것 같아요

kyh196201 replied:
각 route 모두 소문자로 수정했습니다!

passwd10 reviewed (commented):
> 레스토랑 상세 페이지 경로를 설정하는게 많이 고민됐습니다.
결국 api url이랑 동일하게 /restaurants/restaurantId로 설정했는데,
리액트 라우터에서 nested route를 사용할 때 이런식으로 많이 표현하더라구요.
저는 상세 페이지를 /restaurants의 자식 라우터로 설정하지 않았는데, 잘못된 url 설정인가요??

`상세페이지를 /restaurants 의 자식라우터로 설정하지 않았다`는 의미가, RestuarantsPage 내부에서 라우터를 추가로 선언해서 RestaurantdetailPage 를 등록하겠다는 말씀이실까요?

공통으로 사용하는 ui가 없다면 굳이 컴포넌트까지 계층을 구분하지 않아도 될 것 같아요.
만약에 계층을 구분해서 RestuarantsPage 하위에서 라우터를 통해 RestaurantdetailPage 를 등록한다면
RestuarantsPage 컴포넌트가 라우터의 역할도 추가로 해야한다는 문제가 있을 것 같아요. 또 라우터가 여기저기 산재되어있으면 url을 한번에 파악하기도 어려울 것 같아요!

> 최초에 접근했을 때 로딩 메시지가 나오기 위해서 loading의 초기 값을 true로 설정하는게 좋을지
아니면 loadRestaurant 액션에서 loading의 값을 true로 변경해주는게 좋을지 궁금합니다.

말씀하신 loading 상태가 페이지의 컴포넌트들이 모두 렌더링되는것을 기다리기 위함일까요, 단순히 fetching을 통해 데이터를 불러오는 중임을 나타내기 위함인가요? loading의 역할에 따라 초기값을 true로 줄지 action에서 처리할지 달라질 것 같아요.

fetching 용 loading 과 페이지 렌더링용 loading을 둘다 선언해보시고 어느시점에서 어떻게 컨트롤할지 고민해보시면 좋을 것 같아요 ㅎㅎ

kyh196201 committed:
RestaurantDetailContainer 리팩토링

- menu를 렌더링하는 코드의 가독성이 좋지 않아서 삼항 연산자를 사용하지 않도록
수정했습니다.

kyh196201 committed:
capitalizeFirstLetter 유틸 함수 추가

- 단어의 첫 글자를 대문자로 변경하는 함수를 추가했습니다.

kyh196201 committed:
HomePage 리팩토링

- 링크 href가 대문자로 시작하고 있었는데, 소문자로 변경했습니다.

kyh196201 committed:
msw 실행 코드 제거

- api가 정상적으로 동작해서 msw를 실행하는 코드를 제거했습니다.

kyh196201 committed:
restaurant fixture 데이터 수정

- 실제 api에서 내려주는 데이터와 동일한 구조로 변경했습니다.

kyh196201 committed:
RestaurantDetailContainer 리팩토링

- 레스토랑 데이터 구조가 변경됨에 따라 menu를 렌더링하는 코드를 변경했습니다.

kyh196201 committed:
renderRestaurantDetailPage 테스트 리팩토링

- 레스토랑 데이터 구조가 변경됨에따라 테스트 코드를 수정했습니다.

kyh196201 committed:
RestaurantDetailContainer 리팩토링

- 레스토랑이 없을 경우에 대한 테스트를 추가했습니다.

- 레스토랑이 없을 경우 조회 실패 메세지를 렌더링하도록 수정했습니다.

kyh196201 committed:
loadRestaurant 액션 및 테스트 추가

- 레스토랑 상세 정보를 불러오는 비동기 액션 loadRestaurant을 추가했습니다.

kyh196201 committed:
setIsFetchingDetail 액션 오타 수정

kyh196201 committed:
reducer setRestaurant, setIsFetchingDetail 액션 추가

kyh196201 committed:
fetchRestaurant api 테스트 추가

kyh196201 committed:
RestaurantDetailPage 최초 진입 시 레스토랑 상세 API 요청

kyh196201 committed:
loadRestaurant 액션 파라미터 수정

- restaurantId 파라미터를 객체 형태로 받도록 수정했습니다.

kyh196201 committed:
RestaurantDetailPage useParams 적용

- url 경로에서 레스토랑 id를 가져오기 위해서 useParams를 적용했습니다.

kyh196201 committed:
isEmptyObject util 함수 추가

- 빈 객체인지 확인하는 유틸 함수와 테스트를 추가했습니다.

kyh196201 committed:
RestaurantDetail 컴포넌트 추가

kyh196201 committed:
RestaurantDetail 테스트 리팩토링

- 레스토랑 조회에 성공했는데, 조회한 데이터가 없을 경우에 대한 테스트의 제목을 수정했습니다.

kyh196201 committed:
RestaurantDetailContainer 리팩토링

- UI에 관련된 부분을 RestaurantDetail 컴포넌트로 대체했습니다.

kyh196201 committed:
store initialState 수정

- loading을 제거하고 isFetchingDetail을 추가했습니다.

kyh196201 committed:
NotFoundPage 문구 수정

-e2e 테스트 통과를 위해서 메시지를 수정했습니다.

kyh196201 committed:
App 테스트 수정

- NotFoundPage에 대한 테스트를 수정했습니다.

kyh196201 committed:
App 라우터 구조 변경

kyh196201 committed:
Header 링크 적용

- Header 로고 텍스트에 메인 페이지로 이동하는 링크를 설정했습니다.

kyh196201 committed:
index.html main.js 파일 경로 수정

kyh196201 committed:
package.json 마지막 행에 개행 문자 추가

kyh196201 commented:
트레이너님 9시가 넘었지만 부족한 부분이 있어서 코드를 더 수정했습니다 😂
리뷰 부탁드립니다.

## 작업 내용

1. 레스토랑 상세 페이지 구현
2. App 컴포넌트 라우트 구조 변경
3. 피드백 주신 내용 수정

---

이전에 피드백 주신 내용에 대해 공부한 내용입니다.

### 1. package.json에 `No newline at end of file(EOF)`가 나오는 이유가 무엇인지에 대해서 찾아봤습니다.

`POSIX`에서 정한 표준에 따라 파일의 마지막 행에 개행 문자를 포함하는 것을 권장하기 때문입니다.
특히, 특정 프로그램에서 개행 문자 없이 종료된 파일의 마지막 줄을 처리하는 것이 어려울 수 있는데, 이 때문에 파일의 마지막 줄에 개행 문자가 포함되는 것을 권장하고 있습니다.

### 2. 라우터 구조 관련

[이 리뷰](https://github.com/CodeSoom/react-week6-assignment-1/pull/98#discussion_r922278277)에서 언급해주신 구조에 대해서 이해했습니다 😃
하지만 이렇게 작성할 경우 `RestaurantsPage` 내부에 `RestaurantsDetailPage`이 렌더링되기 때문에 현재 과제 구조와 맞지 않다고 생각합니다.
따라서, 현재 라우터 구조를 유지하면서 계층 구조를 좀더 잘 표현하기 위해서 라우터 구조를 수정했습니다.
(라우터 구조는 [2scent님의 구조를 참고했습니다.](https://github.com/CodeSoom/react-week6-assignment-1/pull/97/files#diff-d274a54187c91ba0f532df2a9e194e27ab50e988f5e4c33f5a7893918320c661R24))

---

passwd10 commented on src/containers/RestaurantDetailContainer.test.jsx:
```diff
@@ -32,6 +32,16 @@ describe('<RestaurantDetailContainer />', () => {
+  context('without restaurant', () => {
+    given('restaurant', () => null);
+
+    it('renders "레스토랑을 조회하지 못했습니다." message', () => {
```
안내메세지가 수정되면 테스트도 수정되어야할거에요. 렌더링대상을 범용적으로 표현하면 좋을 것 같아요!

passwd10 commented on src/components/RestaurantDetail.jsx:
```diff
@@ -0,0 +1,43 @@
+import { isEmptyObject } from '../utils';
+
+export default function RestaurantDetail({ restaurant, isFetching }) {
+  if (isFetching || !restaurant) {
+    return (
+      <div>loading...</div>
+    );
+  }
+
+  if (isEmptyObject(restaurant)) {
+    return (
+      <div>
+        레스토랑 정보 조회에 실패했습니다.
+      </div>
+    );
+  }
```
일반적으로 렌더링에 대한 분기처리는 최하단에 모아서 해주는편이에요.

컴포넌트가 무엇을 렌더링할지 한번에 파악하기 편하다는점도 있고, 변수선언이나 로직처리가 다 끝나고 렌더링하게되기 때문에 예상치 못한 사이드이펙트를 방지할 수 있어요.
